math/dec.go

```diff
@@ -1,969 +1,288 @@
-// Would panic if the conversion failed.
-func (d LegacyDec) MustFloat64() float64 {
-	if value, err := strconv.ParseFloat(d.String(), 64); err != nil {
-		panic(err)
-	} else {
-		return value
-	}
+// NewDecFinite returns a decimal with a value of coeff * 10^exp.
+func NewDecFinite(coeff int64, exp int32) Dec {
+	var res Dec
+	res.dec.SetFinite(coeff, exp)
+	return res
 }
 
-//     ____
-//  __|    |__   "chop 'em
-//       ` \     round!"
-// ___||  ~  _     -bankers
-// |         |      __
-// |       | |   __|__|__
-// |_____:  /   | $$$    |
-//              |________|
-
-// Remove a Precision amount of rightmost digits and perform bankers rounding
-// on the remainder (gaussian rounding) on the digits which have been removed.
-//
-// Mutates the input. Use the non-mutative version if that is undesired
-func chopPrecisionAndRound(d *big.Int) *big.Int {
-	// remove the negative and add it back when returning
-	if d.Sign() == -1 {
-		// make d positive, compute chopped value, and then un-mutate d
-		d = d.Neg(d)
-		d = chopPrecisionAndRound(d)
-		d = d.Neg(d)
-		return d
-	}
-
-	// get the truncated quotient and remainder
-	quo, rem := d, big.NewInt(0)
-	quo, rem = quo.QuoRem(d, precisionReuse, rem)
-
-	if rem.Sign() == 0 { // remainder is zero
-		return quo
-	}
-
-	switch rem.Cmp(fivePrecision) {
-	case -1:
-		return quo
-	case 1:
-		return quo.Add(quo, oneInt)
-	default: // bankers rounding must take place
-		// always round to an even number
-		if quo.Bit(0) == 0 {
-			return quo
-		}
-		return quo.Add(quo, oneInt)
-	}
+// Add returns a new Dec with value `x+y` without mutating any argument and error if
+// there is an overflow.
+func (x Dec) Add(y Dec) (Dec, error) {
+	var z Dec
+	_, err := apd.BaseContext.Add(&z.dec, &x.dec, &y.dec)
+	return z, errors.Wrap(err, "decimal addition error")
 }
 
-func chopPrecisionAndRoundUp(d *big.Int) *big.Int {
-	// remove the negative and add it back when returning
-	if d.Sign() == -1 {
-		// make d positive, compute chopped value, and then un-mutate d
-		d = d.Neg(d)
-		// truncate since d is negative...
-		chopPrecisionAndTruncate(d)
-		d = d.Neg(d)
-		return d
-	}
-
-	// get the truncated quotient and remainder
-	quo, rem := d, big.NewInt(0)
-	quo, rem = quo.QuoRem(d, precisionReuse, rem)
-
-	if rem.Sign() == 0 { // remainder is zero
-		return quo
-	}
-
-	return quo.Add(quo, oneInt)
+// Sub returns a new Dec with value `x-y` without mutating any argument and error if
+// there is an overflow.
+func (x Dec) Sub(y Dec) (Dec, error) {
+	var z Dec
+	_, err := apd.BaseContext.Sub(&z.dec, &x.dec, &y.dec)
+	return z, errors.Wrap(err, "decimal subtraction error")
 }
 
-func chopPrecisionAndRoundNonMutative(d *big.Int) *big.Int {
-	tmp := new(big.Int).Set(d)
-	return chopPrecisionAndRound(tmp)
+// Quo returns a new Dec with value `x/y` (formatted as decimal128, 34 digit precision) without mutating any
+// argument and error if there is an overflow.
+func (x Dec) Quo(y Dec) (Dec, error) {
```

Comment:
I did some tests with Quo and the precision is much higher than 34 digits. Do you need to do rounding here?
https://gist.github.com/alpe/5a27babec8005fa347fdd0f0094b1a75

Reply:
Maybe, we should make sure we agree on the expected behavior and then have tests to check it.